File: mobile-android/app/src/main/java/com/example/p2paudio/protocol/QrPayloadCodec.kt
package com.example.p2paudio.protocol

import com.example.p2paudio.model.PairingConfirmPayload
import com.example.p2paudio.model.PairingInitPayload
import java.io.ByteArrayOutputStream
import java.util.Base64
import java.util.zip.Deflater
import java.util.zip.Inflater
import kotlinx.serialization.json.Json

object QrPayloadCodec {
    private val json = Json {
        ignoreUnknownKeys = false
        encodeDefaults = true
        prettyPrint = false
    }

    fun encodeInit(payload: PairingInitPayload): String {
        val raw = json.encodeToString(PairingInitPayload.serializer(), payload)
        return encodeTransportString(raw)
    }

    fun encodeConfirm(payload: PairingConfirmPayload): String {
        val raw = json.encodeToString(PairingConfirmPayload.serializer(), payload)
        return encodeTransportString(raw)
    }

    fun decodeInit(raw: String): PairingInitPayload {
        val decoded = decodeTransportString(raw)
        return json.decodeFromString(PairingInitPayload.serializer(), decoded)
    }

    fun decodeConfirm(raw: String): PairingConfirmPayload {
        val decoded = decodeTransportString(raw)
        return json.decodeFromString(PairingConfirmPayload.serializer(), decoded)
    }

    private fun encodeTransportString(raw: String): String {
        val utf8 = raw.toByteArray(Charsets.UTF_8)
        if (utf8.size < MIN_BYTES_FOR_COMPRESSION) {
            return raw
        }

        val compressed = compress(utf8) ?: return raw
        val encoded = Base64.getUrlEncoder().withoutPadding().encodeToString(compressed)
        if (encoded.length >= raw.length) {
            return raw
        }
        return "$COMPRESSED_PREFIX$encoded"
    }

    private fun decodeTransportString(raw: String): String {
        if (!raw.startsWith(COMPRESSED_PREFIX)) {
            return raw
        }
        val encoded = raw.removePrefix(COMPRESSED_PREFIX)
        require(encoded.isNotBlank()) { "Compressed payload is empty" }

        val compressed = try {
            Base64.getUrlDecoder().decode(encoded)
        } catch (_: IllegalArgumentException) {
            throw IllegalArgumentException("Compressed payload base64 is invalid")
        }
        val decompressed = decompress(compressed, MAX_DECOMPRESSED_BYTES)
            ?: throw IllegalArgumentException("Failed to decode compressed payload")
        return decompressed.toString(Charsets.UTF_8)
    }

    private fun compress(input: ByteArray): ByteArray? {
        val deflater = Deflater(Deflater.BEST_COMPRESSION)
        return try {
            deflater.setInput(input)
            deflater.finish()
            val output = ByteArrayOutputStream(input.size)
            val buffer = ByteArray(1024)
            while (!deflater.finished()) {
                val count = deflater.deflate(buffer)
                if (count <= 0) {
                    break
                }
                output.write(buffer, 0, count)
            }
            if (!deflater.finished()) {
                null
            } else {
                output.toByteArray()
            }
        } catch (_: Exception) {
            null
        } finally {
            deflater.end()
        }
    }

    private fun decompress(input: ByteArray, maxBytes: Int): ByteArray? {
        val inflater = Inflater()
        return try {
            inflater.setInput(input)
            val output = ByteArrayOutputStream(input.size * 2)
            val buffer = ByteArray(1024)
            var total = 0

            while (!inflater.finished()) {
                val count = inflater.inflate(buffer)
                if (count > 0) {
                    total += count
                    if (total > maxBytes) {
                        return null
                    }
                    output.write(buffer, 0, count)
                    continue
                }
                if (inflater.needsInput() || inflater.needsDictionary()) {
                    return null
                }
            }
            output.toByteArray()
        } catch (_: Exception) {
            null
        } finally {
            inflater.end()
        }
    }

    private const val COMPRESSED_PREFIX = "p2paudio-z1:"
    private const val MIN_BYTES_FOR_COMPRESSION = 256
    private const val MAX_DECOMPRESSED_BYTES = 512_000
}

File: desktop-windows/src/P2PAudio.Windows.Core/Protocol/QrPayloadCodec.cs
using System.IO.Compression;
using System.Text;
using System.Text.Json;
using P2PAudio.Windows.Core.Models;

namespace P2PAudio.Windows.Core.Protocol;

public static class QrPayloadCodec
{
    private const string CompressedPrefix = "p2paudio-z1:";
    private const int MinBytesForCompression = 256;
    private const int MaxDecompressedBytes = 512_000;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = false,
        WriteIndented = false
    };

    public static string EncodeInit(PairingInitPayload payload)
    {
        var raw = JsonSerializer.Serialize(payload, JsonOptions);
        return EncodeTransportString(raw);
    }

    public static string EncodeConfirm(PairingConfirmPayload payload)
    {
        var raw = JsonSerializer.Serialize(payload, JsonOptions);
        return EncodeTransportString(raw);
    }

    public static PairingInitPayload DecodeInit(string raw)
    {
        var normalized = DecodeTransportString(raw);
        var payload = JsonSerializer.Deserialize<PairingInitPayload>(normalized, JsonOptions);
        return payload ?? throw new SessionFailure(FailureCode.InvalidPayload, "Invalid init payload");
    }

    public static PairingConfirmPayload DecodeConfirm(string raw)
    {
        var normalized = DecodeTransportString(raw);
        var payload = JsonSerializer.Deserialize<PairingConfirmPayload>(normalized, JsonOptions);
        return payload ?? throw new SessionFailure(FailureCode.InvalidPayload, "Invalid confirm payload");
    }

    private static string EncodeTransportString(string raw)
    {
        var utf8 = Encoding.UTF8.GetBytes(raw);
        if (utf8.Length < MinBytesForCompression)
        {
            return raw;
        }

        var compressed = Compress(utf8);
        if (compressed is null)
        {
            return raw;
        }
        var encoded = ToBase64Url(compressed);
        if (encoded.Length >= raw.Length)
        {
            return raw;
        }
        return $"{CompressedPrefix}{encoded}";
    }

    private static string DecodeTransportString(string raw)
    {
        if (!raw.StartsWith(CompressedPrefix, StringComparison.Ordinal))
        {
            return raw;
        }

        var encoded = raw[CompressedPrefix.Length..];
        if (string.IsNullOrWhiteSpace(encoded))
        {
            throw new SessionFailure(FailureCode.InvalidPayload, "Compressed payload is empty");
        }

        byte[] compressed;
        try
        {
            compressed = FromBase64Url(encoded);
        }
        catch (FormatException)
        {
            throw new SessionFailure(FailureCode.InvalidPayload, "Compressed payload base64 is invalid");
        }

        var decompressed = Decompress(compressed, MaxDecompressedBytes);
        if (decompressed is null)
        {
            throw new SessionFailure(FailureCode.InvalidPayload, "Failed to decode compressed payload");
        }
        return Encoding.UTF8.GetString(decompressed);
    }

    private static byte[]? Compress(byte[] input)
    {
        try
        {
            using var output = new MemoryStream();
            using (var compressor = new ZLibStream(output, CompressionLevel.Fastest, leaveOpen: true))
            {
                compressor.Write(input, 0, input.Length);
            }
            return output.ToArray();
        }
        catch
        {
            return null;
        }
    }

    private static byte[]? Decompress(byte[] input, int maxBytes)
    {
        try
        {
            using var source = new MemoryStream(input);
            using var inflater = new ZLibStream(source, CompressionMode.Decompress);
            using var output = new MemoryStream(input.Length * 2);
            var buffer = new byte[1024];
            while (true)
            {
                var read = inflater.Read(buffer, 0, buffer.Length);
                if (read <= 0)
                {
                    break;
                }
                output.Write(buffer, 0, read);
                if (output.Length > maxBytes)
                {
                    return null;
                }
            }
            return output.ToArray();
        }
        catch
        {
            return null;
        }
    }

    private static string ToBase64Url(byte[] bytes)
    {
        return Convert.ToBase64String(bytes)
            .TrimEnd('=')
            .Replace('+', '-')
            .Replace('/', '_');
    }

    private static byte[] FromBase64Url(string encoded)
    {
        var normalized = encoded
            .Replace('-', '+')
            .Replace('_', '/');
        var remainder = normalized.Length % 4;
        if (remainder > 0)
        {
            normalized = normalized.PadRight(normalized.Length + (4 - remainder), '=');
        }
        return Convert.FromBase64String(normalized);
    }
}
